File: Template/Template.go
package main

import "fmt"

type Game interface {
	initialize();
	startPlay();
	endPlay();
}

func play(g Game){
	g.initialize()
	g.startPlay()
	g.endPlay()
}


type Cricket struct {

}

func(c *Cricket)initialize(){
	fmt.Println("Cricket Game Initialized! Start playing.");
}

func(c *Cricket)startPlay(){
	fmt.Println("Cricket Game Started. Enjoy the game!");
}

func(c *Cricket)endPlay(){
	fmt.Println("Cricket Game Finished!");
}

type Football struct {

}

func(c *Football)initialize(){
	fmt.Println("Football Game Initialized! Start playing.");
}

func(c *Football)startPlay(){
	fmt.Println("Football Game Started. Enjoy the game!");
}

func(c *Football)endPlay(){
	fmt.Println("Football Game Finished!");
}

func main(){
	//demo
	cricket:=new(Cricket)
	play(cricket)

	football:=new(Football)
	play(football)
}

File: Template/Template.php
<?php
/**
 * Created by PhpStorm.
 * User: jamesyan
 * Date: 2017/3/25
 * Time: 19:09
 */

abstract class Game{
    abstract  function initialize();
    abstract  function startPlay();
    abstract  function endPlay();

    public final function play(){
        //初始化
        $this->initialize();
        //开始游戏
        $this->startPlay();
        //结束游戏
        $this->endPlay();
    }
}


class Cricket extends Game{
    function initialize()
    {
        // TODO: Implement initialize() method.
        var_dump("Cricket Game Initialized! Start playing.");
    }

    function startPlay()
    {
        // TODO: Implement startPlay() method.
        var_dump("Cricket Game Started. Enjoy the game!");
    }

    function endPlay()
    {
        // TODO: Implement endPlay() method.
        var_dump("Cricket Game Finished!");
    }
}



class Football extends Game{
    function initialize()
    {
        // TODO: Implement initialize() method.
        var_dump("Football Game Initialized! Start playing.");
    }

    function startPlay()
    {
        // TODO: Implement startPlay() method.
        var_dump("Football Game Started. Enjoy the game!");
    }

    function endPlay()
    {
        // TODO: Implement endPlay() method.
        var_dump("Football Game Finished!");
    }
}


//demo
$game=new Cricket();
$game->play();

$game=new Football();
$game->play();

